pkg/dockerregistry/server/pullthroughmanifestservice.go

```diff
@@ -0,0 +1,83 @@
+package server
+
+import (
+	"github.com/docker/distribution"
+	"github.com/docker/distribution/context"
+	"github.com/docker/distribution/digest"
+
+	imageapi "github.com/openshift/origin/pkg/image/api"
+)
+
+// pullthroughManifestService wraps a distribution.ManifestService
+// repositories. Since the manifest is no longer stored in the Image
+// the docker-registry must pull through requests to manifests as well
+// as to blobs.
+type pullthroughManifestService struct {
+	distribution.ManifestService
+
+	repo                       *repository
+	pullFromInsecureRegistries bool
+}
+
+var _ distribution.ManifestService = &pullthroughManifestService{}
+
+func (m *pullthroughManifestService) Get(ctx context.Context, dgst digest.Digest, options ...distribution.ManifestServiceOption) (distribution.Manifest, error) {
+	manifest, err := m.ManifestService.Get(ctx, dgst, options...)
+	switch err.(type) {
+	case distribution.ErrManifestUnknownRevision:
+		break
+	case nil:
+		return manifest, nil
+	default:
+		return nil, err
+	}
+
+	return m.remoteGet(ctx, dgst, options...)
+}
+
+func (m *pullthroughManifestService) remoteGet(ctx context.Context, dgst digest.Digest, options ...distribution.ManifestServiceOption) (distribution.Manifest, error) {
```

Comment:
Is there a test case that covers manifest pull through?  If not, can you add an extended test for it?